(Rust implementation) Add idiomatic rust iterator for Vector

jaynus opened the pull request:
Implemented for T: Follow (which provides get accessor). 

In rust, it is more idiomatic to use iterators and combinators on 'collections' of data instead of a len()/get() set. The Flatbuffers implementation currently requires:

```for i in 0..vector.len() {

}```

This PR allows more rust-ish:
```for obj in vector.iter() {
}

let transformed = vector.iter.map(|i| i.id()).collect::Vec<u32>(); // psuedo-example

```

Immutable iterator that respects lifetimes. This allows for more idiomatic combinator/iterator access of vector types in flatbuffers. Iterator respects inner vector lifetimes, and allocates single iterator object for tracking (reference + iterator location track).

rw commented:
@jaynus LGTM! Would you add a no-op commit to get CI to run again? Looks like Appveyor had an error.

jaynus commented:
Done and passed.

rw commented:
@jaynus Would you please add a test for this? Perhaps as a sub-`mod` under the `roundtrip_vectors` module in the test suite? Link: https://github.com/google/flatbuffers/blob/master/tests/rust_usage_test/tests/integration_test.rs#L774

rw commented:
@jaynus Or, minimally, by adding a vector iterator assert to the `serialized_example_is_accessible_and_correct` function. Link: https://github.com/google/flatbuffers/blob/master/tests/rust_usage_test/tests/integration_test.rs#L185

tymcauley commented:
This would be a great change! In [this commit](https://github.com/google/flatbuffers/commit/9e82ee25275e07e35f489f6feac543611f81fb83) that was just merged, I had to add some code in a few spots to `tests/rust_usage_test/tests/integration_test.rs` that looks like this:

```
let mut rust_vec: Vec<T> = Vec::with_capacity(flatbuffer_vector.len());
for i in flatbuffer_vector.len() {
    rust_vec.push(flatbuffer_vector.get(i));
}
```

Those might be good candidates to replace with code that tests out this new functionality. Not exactly sure, but maybe something like this:

```
let rust_vec = flatbuffer_vector.iter().collect::Vec<T>();
```

jaynus commented:
As per the commit note, added a small test addition to roundtrip_vectors, which is implemented in the prop function which creates and tests a vector of a given type. this means the iterator is tested for every prop generation test cycle (every time used in roundtrip vectors)

rw commented:
@jaynus Cool! Let's add your vector `collect` check to some more tests, so we can test more situations:

`vector_of_string_store_helper_build`
`vector_of_string_store_manual_build`
`vector_of_f64_store`
`vector_of_table_store`
`vector_of_bool_store`
`vector_of_struct_store`

rw commented on rust/flatbuffers/src/vector.rs:
```diff
@@ -52,6 +52,9 @@ impl<'a, T: Follow<'a> + 'a> Vector<'a, T> {
         T::follow(self.0, self.1 as usize + SIZE_UOFFSET + sz * idx)
     }
+
+    #[inline(always)]
```
Let's just use `#[inline]` here -- it's more idiomatic and gives the compiler more flexibility.

jaynus replied:
Understood.

TethysSvensson commented:
I don't think this works as expected with regards to lifetimes.

Specifically, the items returned by the iterator has a lifetime limited by the **borrowing** of the vector, instead of by the lifetime argument in the vector itself.

Specifically, I would expect all four of these functions to compile (and give the same result):

https://gist.github.com/TethysSvensson/2615a8a1aff7f243fda78c3c1e1ced51

In the current implementation, only the first and last one compiles.

I think part of it is that `Vector` does not implement `Copy` and `Clone` even though it easily could.

TethysSvensson commented:
I think this PR should include something along the lines of this to work as expected:

https://github.com/TethysSvensson/flatbuffers/commit/cb7d088f346dd5341b72a89daba2156465a20070

(This commit comes from a branch we use at my work. It contains a cherry-pick of your commit, but also adds additional derives to the generated enums that we don't expect to get upstreamed.)

rw commented:
@TethysSvensson Thanks for thinking hard about the lifetimes! @jaynus WDYT?

jaynus commented:
I've updated this PR with multiple cases. 

- took the recommended commit of correct lifetimes and implementations of clone+copy, as these make sense as they are just referential structures.
- Added all the requested integration tests

rw commented on rust/flatbuffers/src/vector.rs:
```diff
@@ -28,6 +28,14 @@ use primitives::*;
+    }
+}
+
+impl<'a, T: Follow<'a> + 'a> Copy for Vector<'a, T> {}
```
@jaynus WDYT about deriving this, instead? (May also be possible for `clone`.)

jean-airoldie commented:
Ping

jean-airoldie commented:
@rw Deriving Clone an Copy is not possible in this case since it puts a Copy / Clone bound on T.

rw commented:
@jean-airoldie @jaynus Is this ready to go in? I know there was the other PR @jean-airoldie worked on, so I want to make sure nothing is missed. Otherwise this LGTM.

jean-airoldie commented:
@rw Yes its pretty much ready. The only thing I found is that some bound could be [less strict](https://github.com/google/flatbuffers/pull/5393/files#diff-95ff4721d1c72fd4d40a5ef98182d248R33). But if you merge this PR I'll make another one to fix that.

jaynus commented:
I am here. This has just fallen through the cracks of my github alerts. What is needed to get this finalized and merged?

jean-airoldie commented on rust/flatbuffers/src/vector.rs:
```diff
@@ -28,6 +28,14 @@ use primitives::*;
 #[derive(Debug)]
 pub struct Vector<'a, T: 'a>(&'a [u8], usize, PhantomData<T>);
 
+impl<'a, T: Follow<'a> + 'a> Clone for Vector<'a, T> {
```
This follow bound does not seem necessary.

jean-airoldie commented on rust/flatbuffers/src/vector.rs:
```diff
@@ -135,3 +146,45 @@ impl<'a, T: Follow<'a> + 'a> Follow<'a> for Vector<'a, T> {
+    }
+}
+
+impl<'a, T: Follow<'a> + 'a> Clone for VectorIter<'a, T> {
```
This follow bound does not seem necessary.

jean-airoldie commented on rust/flatbuffers/src/vector.rs:
```diff
@@ -28,6 +28,14 @@ use primitives::*;
+    }
+}
+
+impl<'a, T: Follow<'a> + 'a> Copy for Vector<'a, T> {}
```
This follow bound does not seem necessary.

jaynus replied:
Done.

jean-airoldie commented:
LGTM

jean-airoldie commented on rust/flatbuffers/src/vector.rs:
```diff
@@ -44,14 +52,17 @@ impl<'a, T: 'a> Vector<'a, T> {
 }
 
-impl<'a, T: Follow<'a> + 'a> Vector<'a, T> {
+impl<'a, T: 'a> Vector<'a, T> {
```
Missing Follow bound here

jaynus replied:
Swapped it back.

jean-airoldie commented on rust/flatbuffers/src/vector.rs:
```diff
@@ -52,7 +52,7 @@ impl<'a, T: 'a> Vector<'a, T> {
 }
 
-impl<'a, T: 'a> Vector<'a, T> {
+impl<'a, T: 'a> Vector<'a, T: Follow<'a> + 'a> {
```
Should be ` impl<'a, T: 'a + Follow<'a>> Vector<'a, T> {`

jaynus replied:
Woops What I got for editing on the road :)

TethysSvensson commented on rust/flatbuffers/src/vector.rs:
```diff
@@ -28,6 +28,14 @@ use primitives::*;
 
+impl<'a, T: 'a> Clone for Vector<'a, T> {
+    fn clone(&self) -> Self {
+        Vector(self.0, self.1, self.2)
```
You can use the `Copy` instance to implement the `Clone` instance. This makes it more clear that `Copy == Clone` for this type.

Specifically you can replace the body with `*self`.

However it might be late enough in the process that this should be changed in a later commit.

jean-airoldie commented on rust/flatbuffers/src/vector.rs:
```diff
@@ -135,3 +146,45 @@ impl<'a, T: Follow<'a> + 'a> Follow<'a> for Vector<'a, T> {
+    }
+}
+
+impl<'a, T: Follow<'a> + 'a> Iterator for VectorIter<'a, T> {
```
@jaynus We should provide a [`size_hint`] using vector's [`len`] method.

[`size_hint`]: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.size_hint
[`len`]: https://docs.rs/flatbuffers/0.6.0/flatbuffers/struct.Vector.html#method.len

jean-airoldie commented:
@jaynus Ping.

rw commented:
@jaynus @jean-airoldie @TethysSvensson Looks like both PRs went stagnant. What do you think about merging one of them?

TethysSvensson commented:
@rw: I have been managing a fork of this repo for my work and my fork includes a more efficient version of this PR.

I would be happy if I could get some or all of it upstreamed.

So far I have created #5577, #5579 and #5580 which contains most of the changes in my fork.

My version of this PR is in #5579.

rw commented:
Implemented by https://github.com/google/flatbuffers/pull/5579. Thanks @jaynus @jean-airoldie @TethysSvensson for working on this!
